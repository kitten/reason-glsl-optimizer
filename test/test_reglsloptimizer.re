open Reglsloptimizer;

Printexc.record_backtrace(true);

let vs = {|
  #version 300 es

  in vec3 position;
  in vec3 color;

  uniform mat4 mat;

  out vec4 out_color;

  void main() {
    gl_Position = mat * vec4(position.xyz, 1.0);
    out_color = vec4(color.xyz, 1.0);
  }
|};

let fs = {|
  #version 300 es

  in vec4 out_color;
  out vec4 frag_color;

  void main() {
    frag_color = out_color;
  }
|};

let vsMetal = convertShader(~target=Metal, Vertex, vs);
let fsMetal = convertShader(~target=Metal, Fragment, fs);

let vsGL = convertShader(~target=OpenGL, Vertex, vs);
let fsGL = convertShader(~target=OpenGL, Fragment, fs);

print_endline("---vs metal:");
print_endline(getOutput(vsMetal));
print_endline("---vs metal inputs:");
describeInputs(vsMetal) |> Array.iter(desc => print_endline(desc.name));
print_endline("---vs metal uniforms:");
describeUniforms(vsMetal) |> Array.iter(desc => print_endline(desc.name));
print_endline("---fs metal:");
print_endline(getOutput(fsMetal));
print_endline("---fs metal inputs:");
describeInputs(fsMetal) |> Array.iter(desc => print_endline(desc.name));
print_endline("---fs metal uniforms:");
describeUniforms(fsMetal) |> Array.iter(desc => print_endline(desc.name));

print_endline("---vs gl:");
print_endline(getOutput(vsGL));
print_endline("---fs gl inputs:");
describeInputs(vsGL) |> Array.iter(desc => print_endline(desc.name));
print_endline("---fs gl uniforms:");
describeUniforms(vsGL) |> Array.iter(desc => print_endline(desc.name));
print_endline("---fs gl:");
print_endline(getOutput(fsGL));
print_endline("---fs gl inputs:");
describeInputs(fsGL) |> Array.iter(desc => print_endline(desc.name));
print_endline("---fs gl uniforms:");
describeUniforms(fsGL) |> Array.iter(desc => print_endline(desc.name));
