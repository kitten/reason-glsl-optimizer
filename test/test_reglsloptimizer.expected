---vs metal:
#include <metal_stdlib>
#pragma clang diagnostic ignored "-Wparentheses-equality"
using namespace metal;
struct xlatMtlShaderInput {
  float3 position [[attribute(0)]];
  float3 color [[attribute(1)]];
};
struct xlatMtlShaderOutput {
  float4 gl_Position [[position]];
  float4 out_color;
};
struct xlatMtlShaderUniform {
  float4x4 mat;
};
vertex xlatMtlShaderOutput xlatMtlMain (xlatMtlShaderInput _mtl_i [[stage_in]], constant xlatMtlShaderUniform& _mtl_u [[buffer(0)]])
{
  xlatMtlShaderOutput _mtl_o;
  float4 tmpvar_1 = 0;
  tmpvar_1.w = 1.0;
  tmpvar_1.xyz = _mtl_i.position;
  _mtl_o.gl_Position = (_mtl_u.mat * tmpvar_1);
  float4 tmpvar_2 = 0;
  tmpvar_2.w = 1.0;
  tmpvar_2.xyz = _mtl_i.color;
  _mtl_o.out_color = tmpvar_2;
  return _mtl_o;
}


---vs metal inputs:
position
color
---vs metal uniforms:
mat
---fs metal:
#include <metal_stdlib>
#pragma clang diagnostic ignored "-Wparentheses-equality"
using namespace metal;
struct xlatMtlShaderInput {
  float4 out_color;
};
struct xlatMtlShaderOutput {
  float4 frag_color [[color(0)]];
};
struct xlatMtlShaderUniform {
};
fragment xlatMtlShaderOutput xlatMtlMain (xlatMtlShaderInput _mtl_i [[stage_in]], constant xlatMtlShaderUniform& _mtl_u [[buffer(0)]])
{
  xlatMtlShaderOutput _mtl_o;
  _mtl_o.frag_color = _mtl_i.out_color;
  return _mtl_o;
}


---fs metal inputs:
out_color
---fs metal uniforms:
---vs gl:
#version 330   
in vec3 position;
in vec3 color;
uniform mat4 mat;
out vec4 out_color;
void main ()
{
  vec4 tmpvar_1;
  tmpvar_1.w = 1.0;
  tmpvar_1.xyz = position;
  gl_Position = (mat * tmpvar_1);
  vec4 tmpvar_2;
  tmpvar_2.w = 1.0;
  tmpvar_2.xyz = color;
  out_color = tmpvar_2;
}


---fs gl inputs:
position
color
---fs gl uniforms:
mat
---fs gl:
#version 330   
in highp vec4 out_color;
out highp vec4 frag_color;
void main ()
{
  frag_color = out_color;
}


---fs gl inputs:
out_color
---fs gl uniforms:
